packages/composable-controller/src/ComposableController.ts

```diff
@@ -1,29 +1,60 @@
-import { BaseController, BaseControllerV1 } from '@metamask/base-controller';
 import type {
   ActionConstraint,
   BaseConfig,
+  BaseControllerV1,
   BaseState,
   EventConstraint,
   RestrictedControllerMessenger,
   StateConstraint,
   StateMetadata,
   ControllerStateChangeEvent,
 } from '@metamask/base-controller';
+import { BaseController } from '@metamask/base-controller';
+import type { Json, PublicInterface } from '@metamask/utils';
 import type { Patch } from 'immer';
 
 export const controllerName = 'ComposableController';
 
+export const INVALID_CONTROLLER_ERROR =
+  'Invalid controller: controller must have a `messagingSystem` or be a class inheriting from `BaseControllerV1`.';
+
+/**
+ * A universal supertype for the `BaseControllerV1` state object.
+ */
+type ConfigConstraintV1 = BaseConfig & object;
+
+/**
+ * A universal supertype for the `BaseControllerV1` state object.
+ */
+type StateConstraintV1 = BaseState & object;
+
```

Comment:
Curious: why are these `& object` needed?

Reply:
Line 29 essentially translates to "all object types that extend the `BaseState` type."

The reason `object` is used instead of `Record` is because the V1 controller uses the `interface` keyword to define its state and config types, without explicitly specifying an index signature. 

The `Record` type requires an index signature, and is therefore incompatible with an `interface` type by default.